app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java

```diff
@@ -0,0 +1,544 @@
+                    ContentValues contentValues = helper.getNonDataContentValues(persistable);
+                    contentValues.put(DatabaseHelper.DATA_COL, bos.toByteArray());
+                    insertedId = db.insertOrThrow(table, DatabaseHelper.DATA_COL, contentValues);
+                } else {
+                    // store serialized object in file and file pointer in db
+                    File dataFile = HybridFileBackedSqlHelpers.newFileForEntry(dbDir);
+                    ContentValues contentValues = helper.getNonDataContentValues(persistable);
+                    contentValues.put(DatabaseHelper.FILE_COL, dataFile.getAbsolutePath());
+                    byte[] key = generateKeyAndAdd(contentValues);
+                    insertedId = db.insertOrThrow(table, DatabaseHelper.FILE_COL, contentValues);
+
+                    writeStreamToFile(bos, dataFile.getAbsolutePath(), key);
+                }
+            } finally {
+                bos.close();
+            }
+            // won't effect already stored obj id, which is set when reading out of db.
+            // rather, needed in case persistable object is used after being written to storage.
+            persistable.setID((int)insertedId);
+
+            if (insertedId > Integer.MAX_VALUE) {
+                throw new RuntimeException("Waaaaaaaaaay too many values");
+            }
+
+            db.setTransactionSuccessful();
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
+            db.endTransaction();
+        }
+    }
+
+    private ByteArrayOutputStream writeExternalizableToStream(Externalizable extObj) {
+        ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        try {
+            extObj.writeExternal(new DataOutputStream(bos));
+        } catch (IOException e1) {
+            e1.printStackTrace();
+            throw new RuntimeException("Failed to serialize externalizable");
+        }
+        return bos;
+    }
+
+    protected boolean blobFitsInDb(ByteArrayOutputStream blobStream) {
+        return blobStream.size() < ONE_MB_DB_SIZE_LIMIT;
+    }
+
+    protected byte[] generateKeyAndAdd(ContentValues contentValues) {
+        try {
+            byte[] key = CommCareApplication._().createNewSymetricKey().getEncoded();
+            contentValues.put(DatabaseHelper.AES_COL, key);
+            return key;
+        } catch (SessionUnavailableException e) {
+            throw new RuntimeException("Session unavailable; can't generate encryption key.");
+        }
+    }
+
+    private void writeStreamToFile(ByteArrayOutputStream bos, String filename,
+                                   byte[] key) throws IOException {
+        DataOutputStream fileOutputStream = null;
+        try {
+            fileOutputStream = getOutputFileStream(filename, key);
+            bos.writeTo(fileOutputStream);
+        } finally {
+            if (fileOutputStream != null) {
+                fileOutputStream.close();
```

Comment:
My guess is we should wrap this in an IOException wrapper, since we generally swallow stream close IOExceptions. (The method signature is eating it right now)